### PR TITLE
RT-1.12 : BGP always compare MED  OTG test

### PR DESCRIPTION
**Pull request by cprabha**
Hi,

I have updated OTG test for RT-1.12 BGP always compare MED .
Resolves : https://github.com/openconfig/featureprofiles/issues/1774

Thanks,
Prabha

### CODE REVIEW

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+		}
+		for _, ip := range d.Bgp().Ipv6Interfaces().Items() {
+			for _, configPeer := range ip.Peers().Items() {
+				nbrPath := gnmi.OTG().BgpPeer(configPeer.Name())
+				_, ok := gnmi.Watch(t, otg, nbrPath.SessionState().State(), time.Minute, func(val *ygnmi.Value[otgtelemetry.E_BgpPeer_SessionState]) bool {
+					currState, ok := val.Val()
+					return ok && currState.String() == state
+				}).Await(t)
+				if !ok {
+					fptest.LogQuery(t, "BGP reported state", nbrPath.State(), gnmi.Get(t, otg, nbrPath.State()))
```
Comment (greg-dennis) [inline feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:216]: same as above, the first value of Watch contains the last value seen

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+					t.Errorf("No BGP neighbor formed for peer %s", configPeer.Name())
+				}
+			}
+		}
+	}
+}
+
+// verifyBgpTelemetry checks that the dut has an established BGP session with reasonable settings.
+func verifyBgpTelemetry(t *testing.T, dut *ondatra.DUTDevice) {
+	var nbrIP = []string{ateSrc.IPv4, ateDst1.IPv4, ateDst2.IPv4}
```
Comment (greg-dennis) [inline feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:226]: call t.Helper()

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+		if want := oc.Bgp_Neighbor_SessionState_ESTABLISHED; state != want {
+			t.Errorf("BGP peer %s status got %d, want %d", nbr, state, want)
+		}
+	}
+}
+
+// configureOTG configures the interfaces and BGP protocols on an ATE, including
+// advertising some(faked) networks over BGP.
+func configureOTG(t *testing.T, otg *otg.OTG) gosnappi.Config {
+	config := otg.NewConfig(t)
```
Comment (greg-dennis) [inline feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:254]: call t.Helper()

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+			t.Errorf("Traffic Loss Pct for Flow: %s\n got %v, want 0", flowName, lossPct)
+		} else {
+			t.Logf("Traffic Test Passed!")
+		}
+	}
+}
+
+// setMED is used to configure routing policy to set BGP MED on DUT.
+func setMED(t *testing.T, dut *ondatra.DUTDevice, d *oc.Root) {
+	// Configure SetMED on DUT.
```
Comment (greg-dennis) [inline feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:387]: call t.Helper()

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+	// Apply setMed Import policy on eBGP Peer2 - OTG Port3 -  with MED 50.
+	dutPolicyConfPath = gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).
+		Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP").Bgp().
+		PeerGroup(peerGrpName3).ApplyPolicy().ImportPolicy()
+	gnmi.Replace(t, dut, dutPolicyConfPath.Config(), []string{setMEDPolicy50})
+}
+
+// verifySetMed is used to validate MED on received prefixes at OTG Port1.
+func verifySetMed(t *testing.T, otg *otg.OTG, config gosnappi.Config, wantMEDValue uint32) {
+	_, ok := gnmi.WatchAll(t,
```
Comment (greg-dennis) [inline feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:415]: call t.Helper()

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+		otg,
+		gnmi.OTG().BgpPeer(ateSrc.Name+".BGP4.peer").UnicastIpv4PrefixAny().State(),
+		time.Minute,
+		func(v *ygnmi.Value[*otgtelemetry.BgpPeer_UnicastIpv4Prefix]) bool {
+			_, present := v.Val()
```
Comment (greg-dennis) [inline feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:420]: you can just do "return val.IsPresent()"

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+		af4 := nv4.GetOrCreateAfiSafi(oc.BgpTypes_AFI_SAFI_TYPE_IPV4_UNICAST)
+		af4.Enabled = ygot.Bool(true)
+		af6 := nv4.GetOrCreateAfiSafi(oc.BgpTypes_AFI_SAFI_TYPE_IPV6_UNICAST)
+		af6.Enabled = ygot.Bool(false)
+	}
+	return ni_proto
+}
+
+func sendTraffic(t *testing.T, otg *otg.OTG, c gosnappi.Config) {
+	t.Logf("Starting traffic")
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:188]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+	otg.StartTraffic(t)
+	time.Sleep(trafficDuration)
+	t.Logf("Stop traffic")
+	otg.StopTraffic(t)
+}
+
+func verifyOTGBGPTelemetry(t *testing.T, otg *otg.OTG, c gosnappi.Config, state string) {
+	for _, d := range c.Devices().Items() {
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:197]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+		for _, ip := range d.Bgp().Ipv4Interfaces().Items() {
+			for _, configPeer := range ip.Peers().Items() {
+				nbrPath := gnmi.OTG().BgpPeer(configPeer.Name())
+				_, ok := gnmi.Watch(t, otg, nbrPath.SessionState().State(), time.Minute, func(val *ygnmi.Value[otgtelemetry.E_BgpPeer_SessionState]) bool {
+					currState, ok := val.Val()
+					return ok && currState.String() == state
+				}).Await(t)
+				if !ok {
+					fptest.LogQuery(t, "BGP reported state", nbrPath.State(), gnmi.Get(t, otg, nbrPath.State()))
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 1 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 2 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:230]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 5 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:421]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 6 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 10]
```diff
@@ -0,0 +1,583 @@
...
+		otg,
+		gnmi.OTG().BgpPeer(ateSrc.Name+".BGP4.peer").UnicastIpv4PrefixAny().State(),
+		time.Minute,
+		func(v *ygnmi.Value[*otgtelemetry.BgpPeer_UnicastIpv4Prefix]) bool {
+			_, present := v.Val()
+			return present
+		}).Await(t)
+
+	if ok {
+		bgpPrefixes := gnmi.GetAll(t, otg, gnmi.OTG().BgpPeer(ateSrc.Name+".BGP4.peer").UnicastIpv4PrefixAny().State())
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:430]: I did check watch is checking all and at the end will have details of only one entry.  any other  way we can do this ? please let me know

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 11]
```diff
@@ -0,0 +1,583 @@
...
+		}).Await(t)
+		if !ok {
+			t.Errorf("obtained MED on OTG is not as expected")
+		}
+	*/
+}
+
+// verifyBGPCapabilities is used to Verify BGP capabilities like route refresh as32 and mpbgp.
+func verifyBGPCapabilities(t *testing.T, dut *ondatra.DUTDevice) {
+	t.Log("Verifying BGP capabilities.")
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:467]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+				t.Errorf("Capability not reported: %v", cap)
+			}
+		}
+	}
+}
+
+// verifyPrefixesTelemetry confirms that the dut shows the correct numbers of installed,
+// sent and received IPv4 prefixes.
+func verifyPrefixesTelemetry(t *testing.T, dut *ondatra.DUTDevice, wantInstalled, wantSent uint32) {
+	statePath := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP").Bgp()
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:491]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 13]
```diff
@@ -0,0 +1,583 @@
...
+func verifyBgpTelemetry(t *testing.T, dut *ondatra.DUTDevice) {
+	var nbrIP = []string{ateSrc.IPv4, ateDst1.IPv4, ateDst2.IPv4}
+	t.Logf("Verifying BGP state.")
+	bgpPath := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP").Bgp()
+
+	for _, nbr := range nbrIP {
+		nbrPath := bgpPath.Neighbor(nbr)
+		// Get BGP adjacency state.
+		t.Logf("Waiting for BGP neighbor to establish...")
+		var status *ygnmi.Value[oc.E_Bgp_Neighbor_SessionState]
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 14]
```diff
@@ -0,0 +1,583 @@
...
+		status, ok := gnmi.Watch(t, dut, nbrPath.SessionState().State(), time.Minute, func(val *ygnmi.Value[oc.E_Bgp_Neighbor_SessionState]) bool {
+			state, ok := val.Val()
+			return ok && state == oc.Bgp_Neighbor_SessionState_ESTABLISHED
+		}).Await(t)
+		if !ok {
+			fptest.LogQuery(t, "BGP reported state", nbrPath.State(), gnmi.Get(t, dut, nbrPath.State()))
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 3 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:258]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 15]
```diff
@@ -0,0 +1,583 @@
...
+	otg.StartProtocols(t)
+	time.Sleep(30 * time.Second)
+
+	return config
+}
+
+// verifyTraffic confirms that every traffic flow has the expected amount of loss (0% or 100%
+// depending on wantLoss, +- 2%).
+func verifyTraffic(t *testing.T, ate *ondatra.ATEDevice, c gosnappi.Config, flowName string, wantLoss bool) {
+	otg := ate.OTG()
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:367]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 4 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: updated

[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: updated

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 16]
```diff
@@ -0,0 +1,590 @@
...
+	for _, d := range c.Devices().Items() {
+		for _, ip := range d.Bgp().Ipv4Interfaces().Items() {
+			for _, configPeer := range ip.Peers().Items() {
+				nbrPath := gnmi.OTG().BgpPeer(configPeer.Name())
+				status, ok := gnmi.Watch(t, otg, nbrPath.SessionState().State(), time.Minute, func(val *ygnmi.Value[otgtelemetry.E_BgpPeer_SessionState]) bool {
+					currState, ok := val.Val()
+					return ok && currState.String() == state
+				}).Await(t)
+				if !ok {
+					t.Logf("BGP reported state %v", status.String())
```
Comment (greg-dennis) [inline feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:206]: If the logged status is useful, put it in the error message just below and remove the log statement. Also, no need to call .String() on the status, just using %v with the status itself is sufficient.

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 17]
```diff
@@ -0,0 +1,590 @@
...
+		for _, ip := range d.Bgp().Ipv6Interfaces().Items() {
+			for _, configPeer := range ip.Peers().Items() {
+				nbrPath := gnmi.OTG().BgpPeer(configPeer.Name())
+				status, ok := gnmi.Watch(t, otg, nbrPath.SessionState().State(), time.Minute, func(val *ygnmi.Value[otgtelemetry.E_BgpPeer_SessionState]) bool {
+					currState, ok := val.Val()
+					return ok && currState.String() == state
+				}).Await(t)
+				if !ok {
+					t.Logf("BGP reported state %v", status.String())
+					t.Errorf("No BGP neighbor formed for peer %s", configPeer.Name())
```
Comment (greg-dennis) [inline feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:220]: same as above

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 16 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: removed logf message and kept only fatal statement.

File: feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 17 above)
Comment (cprabha) [reply feature/experimental/bgp/otg_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:218]: removed logf here

**Review by greg-dennis — APPROVED** [review]
Approving for Go quality

**Review by greg-dennis — DISMISSED** [review]
Re-approving for Go quality

**Review by greg-dennis — APPROVED** [review]
Re-approving for Go quality

**Pull request merged**